ecosystem/indexer-grpc/indexer-grpc-utils/src/in_memory_cache.rs

```diff
@@ -0,0 +1,439 @@
+            let latest_version = self.latest_version().await;
+            if starting_version >= latest_version {
+                tokio::time::sleep(std::time::Duration::from_millis(
+                    IN_MEMORY_CACHE_LOOKUP_RETRY_INTERVAL_MS,
+                ))
+                .await;
+                continue;
+            }
+            // This is to avoid fetching too many transactions at once.
+            let ending_version = std::cmp::min(
+                latest_version,
+                starting_version + MAX_REDIS_FETCH_BATCH_SIZE as u64,
+            );
+            break (
+                (starting_version..ending_version).collect::<Vec<u64>>(),
+                latest_version,
+            );
+        };
+        let lock_waiting_time = start_time.elapsed().as_secs_f64();
+        let mut arc_transactions = Vec::new();
+        for key in versions_to_fetch {
+            if let Some(transaction) = self.cache.get(&key) {
+                arc_transactions.push(transaction.clone());
+            } else {
+                break;
+            }
+        }
+
+        let map_lookup_time = start_time.elapsed().as_secs_f64();
+        // Actual clone.
+        let res: Vec<Transaction> = arc_transactions
+            .into_iter()
+            .map(|t| t.as_ref().clone())
+            .collect();
+        let actual_copy_time = start_time.elapsed().as_secs_f64();
+        tracing::info!(
+            transactions_count = res.len(),
+            starting_version,
+            in_memory_latest_version,
+            duration_in_seconds = start_time.elapsed().as_secs_f64(),
+            lock_waiting_time,
+            map_lookup_time,
+            actual_copy_time,
+            "In-memory cache lookup",
+        );
+        res
+    }
+}
+
+/// Warm up the cache with the latest transactions.
+async fn warm_up_the_cache<C>(
+    conn: C,
+    cache: Arc<DashMap<u64, Arc<Transaction>>>,
+    storage_format: StorageFormat,
+) -> anyhow::Result<(u64, u64, u64)>
+where
+    C: redis::aio::ConnectionLike + Send + Sync + Clone + 'static,
+{
+    let mut conn = conn.clone();
+    let latest_version = get_config_by_key(&mut conn, "latest_version")
+        .await
+        .context("Failed to fetch the latest version from redis")?
+        .context("Latest version doesn't exist in Redis")?;
+    if latest_version == 0 {
+        return Ok((0, 0, 0));
+    }
+    let versions_to_fetch: Vec<u64> =
+        (latest_version.saturating_sub(WARM_UP_CACHE_ENTRIES)..latest_version).collect();
+    let first_version = versions_to_fetch[0];
+    let transactions = batch_get_transactions(&mut conn, versions_to_fetch, storage_format).await?;
+    let total_size_in_bytes = transactions.iter().map(|t| t.encoded_len() as u64).sum();
+    for transaction in transactions {
+        cache.insert(transaction.version, Arc::new(transaction));
+    }
+    Ok((first_version, latest_version, total_size_in_bytes))
+}
+
+async fn create_update_task<C>(
```

Comment:
does this function need to be async?

Reply:
nit: s/create_update_task/spawn_update_task

Reply:
done. removed async and updated name.